jax/_src/prng.py

```diff
@@ -0,0 +1,358 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from functools import partial
+
+import numpy as np
+
+from jax import lax
+from jax import core
+from jax import numpy as jnp
+from jax import tree_util
+from jax._src.api import jit, vmap
+from jax.lib import xla_bridge
+from jax.lib import xla_client
+from jax.lib import cuda_prng
+from jax.interpreters import batching
+from jax.interpreters import xla
+from jax._src.util import prod
+
+
+UINT_DTYPES = {8: jnp.uint8, 16: jnp.uint16, 32: jnp.uint32, 64: jnp.uint64}
+
+
+@tree_util.register_pytree_node_class
+class PRNGKey:
+  """Represents a PRNG key or batch thereof."""
+
+  key: jnp.ndarray
+
+  def __init__(self, key: jnp.ndarray):
+    # key might be a dummy object due to tree_unflatten
+    ndim = getattr(key, 'ndim', 1)
+    dtype = getattr(key, 'dtype', np.uint32)
+    if ndim < 1 or dtype != np.uint32:
+      raise TypeError(
+          f'invalid prng key or key batch: ndim = {ndim}, dtype = {dtype}')
+    self.key = key
+
+  def tree_flatten(self):
+    return (self.key,), None
+
+  @classmethod
+  def tree_unflatten(cls, _, key):
+    key, = key
+    return cls(key)
+
+  def fold_in(self, data: int) -> 'PRNGKey':
+    return PRNGKey(_fold_in(self.key, data))
+
+  def random_bits(self, bit_width, shape) -> jnp.ndarray:
+    return _random_bits(self.key, bit_width, shape)
+
+  def split(self, num: int) -> 'PRNGKey':
+    return PRNGKey(_split(self.key, num))
+
+  def __iter__(self):
+    assert self.key.ndim > 0
+    if self.key.ndim == 1:
+      raise TypeError('iteration over a single PRNG key')
+    return (PRNGKey(k) for k in self.key.__iter__())
+
+  # TODO(frostig): remove if possible, otherwise declare necessary
+  @property
+  def shape(self):
+    return self.key.shape
+
+  # TODO(frostig): remove if possible, otherwise declare necessary
+  @property
+  def dtype(self):
+    return self.key.dtype
+
+
+def make_prng_key(seed: int) -> PRNGKey:
```

Comment:
Just curious, but would it make more sense for this to be a class factory?
```python
class PRNGKey:
  @classmethod
  def from_seed(cls: Type[T], seed: int) -> T:
    return cls(_threefry_prng_key(seed))
```
This has two possible advantages:

- It would be one fewer thing to expose, and
- If you choose to add it to the abstract RNG base class, it would be part of the interface.